src/met.hxx

```diff
@@ -0,0 +1,389 @@
+ * @param genparticle_pt genparticle pt
+ * @param genparticle_eta genparticle eta
+ * @param genparticle_phi genparticle phi
+ * @param genparticle_mass genparticle mass
+ * @param genparticle_id genparticle PDG ID
+ * @param genparticle_status genparticle status bit
+ * @param jet_pt the pt of all jets
+ * @param outputname name of the new column containing the corrected met
+ * @param recoilfile path to the recoil corrections file
+ * @param systematicsfile path to the systematics corrections file
+ * @param applyRecoilCorrections if bool is set, the recoil correction is
+applied, if not, the outputcolumn contains the original met value
+ * @param resolution bool - if set, resolution corrections are applied
+ * @param response bool - if set, response corrections are applied
+ * @param shiftUp bool - if set, the up shift is applied
+ * @param shiftDown bool - if set, the down shift is applied
+ * @return a new dataframe containing the new met column
+ */
+auto applyRecoilCorrections(
+    auto df, const std::string &met, const std::string &genparticle_pt,
+    const std::string &genparticle_eta, const std::string &genparticle_phi,
+    const std::string &genparticle_mass, const std::string &genparticle_id,
+    const std::string &genparticle_status, const std::string &jet_pt,
+    const std::string &outputname, const std::string &recoilfile,
+    const std::string &systematicsfile, bool applyRecoilCorrections,
+    bool resolution, bool response, bool shiftUp, bool shiftDown) {
+    if (applyRecoilCorrections) {
+        Logger::get("RecoilCorrections")->debug("Will run recoil corrections");
+        const auto corrector = new RecoilCorrector(recoilfile);
+        const auto systematics = new METSystematic(systematicsfile);
+        auto shiftType = METSystematic::SysShift::Nominal;
+        if (shiftUp) {
+            shiftType = METSystematic::SysShift::Up;
+        } else if (shiftDown) {
+            shiftType = METSystematic::SysShift::Down;
+        }
+        auto sysType = METSystematic::SysType::None;
+        if (response) {
+            sysType = METSystematic::SysType::Response;
+        } else if (resolution) {
+            sysType = METSystematic::SysType::Resolution;
+        }
+        auto RecoilCorrections = [sysType, systematics, shiftType, corrector](
+                                     ROOT::Math::PtEtaPhiMVector &met,
+                                     const ROOT::RVec<float> &genparticle_pt,
+                                     const ROOT::RVec<float> &genparticle_eta,
+                                     const ROOT::RVec<float> &genparticle_phi,
+                                     const ROOT::RVec<float> &genparticle_mass,
+                                     const ROOT::RVec<int> &genparticle_id,
+                                     const ROOT::RVec<int> &genparticle_status,
+                                     const ROOT::RVec<float> &jet_pt) {
+            // TODO is this the correct number of jets ?
```

Comment:
I think we need some extra treatment for WJets

Reply:
do we run these omly for WJets or also for other samples ?

For WJets we just add one jet here so 

```
int nJets30 = jets_above30.size() + 1;
```

Reply:
also for DY and Signal afaik

Reply:
ok, then we will need an additional flag for this

Reply:
Is that possible, or do we need two separate producers in the python part ?

Reply:
I think a config parameter which is steered via the sample key would do the job.

Reply:
done